Update ERC-7739: Update ERC-7739 (#687)
ethereum/ERCs: ERCS/erc-7739.md

The `PersonalSign` workflow does not require additional data to be appended to the `signature` passed into `isValidSignature`.

### Support detection

> Smart accounts SHOULD return `bytes4(0x77390001)` for `isValidSignature(0x7739773977397739773977397739773977397739773977397739773977397739, "Any Signature")` to indicate support for this standard.

Reviewer: I'd consider requiring the signature argument to be the empty string instead of accepting anything. Just to reduce the surface where `isValidSignature` behaves different than expected.

```suggestion
Smart accounts SHOULD return `bytes4(0x77390001)` for `isValidSignature(0x7739773977397739773977397739773977397739773977397739773977397739, "")` to indicate support for this standard.
```

The `typedDataSignTypehash` must be constructed on-the-fly on-chain. This is to enforce that the signed contents will be visible in the signature request, by requiring that `contents` be a user defined type. 

> The fields of `eip712Domain` are flattened into the `TypedDataSign` structure to enable compatibility with certain Ethers JS versions.

Reviewer: Can you explain this?

Reviewer: Ok I understand.

```suggestion
The fields of `eip712Domain` are flattened into the `TypedDataSign` structure instead of being included as a field of type `EIP712Domain`, because the type could be different from and conflict with the domain type of the verifying contract.
```

The `bytes1 fields` bitmap and `uint256[] extensions` array in [ERC-5267](./eip-5267.md) have been omitted. Differentiating between an absent field versus a zero field (e.g. `bytes32(0)`) offers no additional security benefits for on-chain defensive rehashing. The `extensions` parameter is a list of EIP numbers used for off-chain signaling.

### `contentsDescription` with implicit and explicit modes

> When the `contents` structure contains nested types, EIP-712 lexicographical sorting can result in the `contentsName` not being positioned exactly at the start of the `contentsType`. As such, we need the explicit mode.

Reviewer 2: I understand the rationale behind these two modes but I feel the ERC should provide a general guideline for when to use each other. So far I've found the main benefit of using implicit mode when possible is that it reduces the calldata costs. If that's the case, I'd suggest to include it in this section

```suggestion
When the `contents` structure contains nested types, EIP-712 lexicographical sorting can result in the `contentsName` not being positioned exactly at the start of the `contentsType`. As such, we need the explicit mode.

Signing UIs may use implicit mode whenever possible to reduce the calldata size. Although the `explicit` mode works as a generalized approach.
```

### Support detection with `isValidSignature`

For easier implementation in modular smart accounts, we have decided to utilize the `isValidSignature` method to return a magic number instead of defining new functions.

### Rejecting `contentsName` beginning with any lowercase 7-bit ASCII character

This recommendation is to keep the standard language agnostic and future-proof. Atomic types such as `uint256` may be named differently in other languages (e.g. `u256`).

## Backwards Compatibility

### Detection of previous draft

> In an earlier draft, we have designated a `supportsNestedTypedDataSign()` function for support detection, which returns `bytes4(0xd620c85a)`.

Reviewer 3: Can you please add something like:

"To ensure false negative detections, it is RECOMMENDED that dApps use `supportsNestedTypedDataSign()` as a fallback detection method in addition to the one described earlier in the spec."

to ensure dApps consider using `supportsNestedTypedDataSign()` and not just ignore it
